### PR TITLE
Updated day3 and day4 solution

### PR DESCRIPTION
**Pull request by leowei01**

### CODE REVIEW

File: src/aoc2020/day3_update.clj [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  (position-s slope-dist)
+  #_=> ([0 0] [1 3] [2 6] [3 9] [4 12] ...)
+
+  ;; get the value of the position[row col]
+  (value-at-pos sample-grid [0 0])
+  #_=> \.
+
+  (value-at-pos sample-grid [1 0])
+  #_=> \#
+
```
Comment (lukaszhall) [inline src/aoc2020/day3_update.clj:89]: Where is the sequence of values corresponding to the positions?

Comment (leowei01) [reply src/aoc2020/day3_update.clj:89]: The sequence is infinite, I need to `take` a part of that and then I can map `value-at-pos` to sequence. I will add that.

File: src/aoc2020/day3_update.clj [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  (:require [aoc2020.util :refer [file->seq]]))
+
+(defn grid-template->grid
+  [grid-template]
+  (mapv (fn [row] (cycle (seq row)))
+        grid-template))
+
+(defn position-s
+  [{slope-right :right
+    slope-down  :down}]
```
Comment (lukaszhall) [inline src/aoc2020/day3_update.clj]: You'll generally want to `{:keys ..` form. 
As you are controlling the map coming in, you should be asking why the names of the keys are mis aligned with meaning

File: src/aoc2020/day4.clj [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        value           (->> (drop-last 2 (seq value))
+                             (apply str)
+                             (parse-long))]
+    (cond
+      (= unit-of-measure "cm") (<= 150 value 193)
+      (= unit-of-measure "in") (<= 59 value 76)
+      :else false)))
+
+(defmethod valid-value? :hcl [_field value]
+  (let [rest-chars     (apply str (rest value))
```
Comment (lukaszhall) [inline src/aoc2020/day4.clj:71]: poor name

File: src/aoc2020/day4.clj [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+       (count)))
+
+
+
+
+
+(defmulti valid-value?
+          (fn [field _value] field))
+
+(defmethod valid-value? :byr [_field value]
```
Comment (lukaszhall) [inline src/aoc2020/day4.clj:47]: Use team clj convention for function layout

File: src/aoc2020/day4.clj [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        (file->passport-str-s "aoc2020/day4/sample-input.txt"))
+      sample-passport-str-s)
+  #_=> ["ecl:gry pid:860033327 eyr:2020 hcl:#fffffd byr:1937 iyr:2017 cid:147 hgt:183cm"
+        "iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884 hcl:#cfa07d byr:1929"
+        "hcl:#ae17e1 iyr:2013 eyr:2024 ecl:brn pid:760753108 byr:1931 hgt:179cm"
+        "hcl:#cfa07d eyr:2025 pid:166559648 iyr:2011 ecl:brn hgt:59in"]
+
+  (do (def passport-str-s
+        (file->passport-str-s "aoc2020/day4/input.txt"))
+      passport-str-s)
```
Comment (lukaszhall) [inline src/aoc2020/day4.clj:109]: no value ?

File: src/aoc2020/day4.clj [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+  ;; parse passport-str to password
+  (parse-passport-str (first sample-passport-str-s))
+  #_=> {:ecl "gry", :pid "860033327", :eyr "2020", :hcl "#fffffd", :byr "1937", :iyr "2017", :cid "147", :hgt "183cm"}
+
+  (do (def sample-passport-s (map parse-passport-str sample-passport-str-s))
+      sample-passport-s)
+  #_=> ({:ecl "gry", :pid "860033327", :eyr "2020", :hcl "#fffffd", :byr "1937", :iyr "2017", :cid "147", :hgt "183cm"}
+        {:iyr "2013", :ecl "amb", :cid "350", :eyr "2023", :pid "028048884", :hcl "#cfa07d", :byr "1929"}
+        {:hcl "#ae17e1", :iyr "2013", :eyr "2024", :ecl "brn", :pid "760753108", :byr "1931", :hgt "179cm"}
+        {:hcl "#cfa07d", :eyr "2025", :pid "166559648", :iyr "2011", :ecl "brn", :hgt "59in"})
```
Comment (lukaszhall) [inline src/aoc2020/day4.clj:137]: Missing demonstration of intermediate validation multimethods

**Pull request merged**